libqtile/command/base.py

```diff
@@ -54,14 +56,85 @@ class CommandException(Exception):
     """Error raised while executing a command"""
 
 
-class CommandObject(metaclass=abc.ABCMeta):
+class _CommandMeta(abc.ABCMeta):
+    def __new__(cls, name, bases, dct):
+        instance = super().__new__(cls, name, bases, dct)
+
+        if hasattr(cls, "_commands"):
+            return instance
+
+        commands = {}
+        cmd_s = set()
+
+        for c in bases + (cls,):
+            for method_name in dct.keys():
+                method = getattr(c, method_name, None)
+
+                if not method:
+                    continue
+
+                if hasattr(method, "_cmd") or method_name in commands:
+                    commands[method_name] = method
+                elif method_name.startswith("cmd_"):
+                    cmd_s.add(method_name)
+                    commands[method_name[4:]] = method
+
+                for mapping in getattr(method, "_mapping", list()):
+                    setattr(cls, mapping, method)
+                    commands[mapping] = method
+
+        if cmd_s:
+            names = " ".join(cmd_s)
+            msg = (
+                f"The use of the 'cmd_' prefix to expose commands via IPC "
+                f"is deprecated. Methods should use the "
+                f"@CommandObject.expose decorator instead. "
+                f"Please update: {names}"
+            )
+            logger.warning(f"Deprecation Warning: {msg}")
+
+        return instance
```

Comment:
I can see what this is trying to do (and the code looks nice).

The obvious issue I can see is that we're not assigning the `commands` object to any class or instance attribute. That being said, if I add a line like `cls._commands = commands`, the object seems to be empty so maybe it's going wrong somewhere else...

Will do some more digging as it feels like this should be the approach we use.

Reply:
Partly my mistake, should have been `instance._commands = commands` not `cls` (`cls` is the `_CommandMeta` class...). Still getting some empty `commands` objects here which I wouldn't expect.